rs/https_outcalls/adapter/src/rpc_server.rs

```diff
@@ -98,8 +113,147 @@ impl CanisterHttp {
+            }
+            (Ok(_), Err(err)) => {
+                debug!(
+                    self.logger,
+                    "SOCKS_PROXY_DL: regular request succeeded, DL request failed with error {}",
+                    err,
+                );
+            }
+            (Err(err), Ok(_)) => {
+                debug!(
+                    self.logger,
+                    "SOCKS_PROXY_DL: DL request succeeded, regular request failed with error {}",
+                    err,
+                );
+            }
+        }
+    }
+
+    // Attemts to load socks the client from the cache. If not present, creates a new socks client and adds it to the cache.
+    fn get_socks_client(
+        &self,
+        address: &str,
+    ) -> Option<Client<HttpsConnector<SocksConnector<HttpConnector>>, OutboundRequestBody>> {
+        let cache_guard = self.cache.upgradable_read();
+
+        if let Some(client) = cache_guard.get(address) {
+            Some(client.clone())
+        } else {
+            let mut cache_guard = RwLockUpgradableReadGuard::upgrade(cache_guard);
+            self.metrics.socks_cache_miss.inc();
+            self.maybe_add_socks_client_for_address(&mut cache_guard, address);
+            match cache_guard.get(address) {
+                Some(client) => Some(client.clone()),
+                None => {
+                    debug!(
+                        self.logger,
+                        "Failed to create SOCKS client for address: {}", address
+                    );
+                    None
+                }
+            }
+        }
+    }
+
+    async fn https_outcall_socks_proxy(
+        &self,
+        socks_proxy_addrs: &[String],
+        request: http::Request<Full<Bytes>>,
+    ) -> Result<http::Response<Incoming>, String> {
+        let mut socks_proxy_addrs = socks_proxy_addrs.to_owned();
+
+        socks_proxy_addrs.shuffle(&mut thread_rng());
+
+        let mut last_error = None;
+
+        let mut tries = 0;
+
+        for socks_proxy_addr in &socks_proxy_addrs {
+            tries += 1;
+            if tries > MAX_SOCKS_PROXY_RETRIES {
+                break;
+            }
+            let next_socks_proxy_addr = socks_proxy_addr.clone();
+
+            let socks_client = self.get_socks_client(&next_socks_proxy_addr);
+
+            if socks_client.is_none() {
+                continue;
+            }
+            let socks_client = socks_client.unwrap();
+
+            match socks_client.request(request.clone()).await {
+                Ok(resp) => {
+                    self.metrics
+                        .socks_connections_attempts
+                        .with_label_values(&[&tries.to_string(), "success"])
+                        .inc();
+                    return Ok(resp);
+                }
+                Err(socks_err) => {
+                    self.metrics
+                        .socks_connections_attempts
+                        .with_label_values(&[&tries.to_string(), "failure"])
```

Comment:
Here I am not sure and maybe @rumenov has a better opinion:
How about we add the SOCKS proxy addr as a label? We expect to have maybe 50 API BNs (more realistic 20). So, that shouldn't increase the metrics too much. It would just help us to see the problems.